launcher/ui/pages/instance/DataPackPage.cpp

```diff
@@ -0,0 +1,342 @@
+    ui->actionsToolbar->insertActionBefore(ui->actionAddItem, ui->actionDownloadItem);
+
+    connect(ui->actionDownloadItem, &QAction::triggered, this, &DataPackPage::downloadDataPacks);
+
+    ui->actionUpdateItem->setToolTip(tr("Try to check or update all selected data packs (all data packs if none are selected)"));
+    connect(ui->actionUpdateItem, &QAction::triggered, this, &DataPackPage::updateDataPacks);
+    ui->actionsToolbar->insertActionBefore(ui->actionAddItem, ui->actionUpdateItem);
+
+    auto updateMenu = new QMenu(this);
+
+    auto update = updateMenu->addAction(ui->actionUpdateItem->text());
+    connect(update, &QAction::triggered, this, &DataPackPage::updateDataPacks);
+
+    updateMenu->addAction(ui->actionResetItemMetadata);
+    connect(ui->actionResetItemMetadata, &QAction::triggered, this, &DataPackPage::deleteDataPackMetadata);
+
+    ui->actionUpdateItem->setMenu(updateMenu);
+
+    ui->actionChangeVersion->setToolTip(tr("Change a data pack's version."));
+    connect(ui->actionChangeVersion, &QAction::triggered, this, &DataPackPage::changeDataPackVersion);
+    ui->actionsToolbar->insertActionAfter(ui->actionUpdateItem, ui->actionChangeVersion);
+}
+
+void DataPackPage::updateFrame(const QModelIndex& current, [[maybe_unused]] const QModelIndex& previous)
+{
+    auto sourceCurrent = m_filterModel->mapToSource(current);
+    int row = sourceCurrent.row();
+    auto& dp = m_model->at(row);
+    ui->frame->updateWithDataPack(dp);
+}
+
+void DataPackPage::downloadDataPacks()
+{
+    if (m_instance->typeName() != "Minecraft")
+        return;  // this is a null instance or a legacy instance
+
+    ResourceDownload::DataPackDownloadDialog mdownload(this, std::static_pointer_cast<DataPackFolderModel>(m_model), m_instance);
+    if (mdownload.exec()) {
+        auto tasks = new ConcurrentTask("Download Data Pack", APPLICATION->settings()->get("NumberOfConcurrentDownloads").toInt());
+        connect(tasks, &Task::failed, [this, tasks](QString reason) {
+            CustomMessageBox::selectable(this, tr("Error"), reason, QMessageBox::Critical)->show();
+            tasks->deleteLater();
+        });
+        connect(tasks, &Task::aborted, [this, tasks]() {
+            CustomMessageBox::selectable(this, tr("Aborted"), tr("Download stopped by user."), QMessageBox::Information)->show();
+            tasks->deleteLater();
+        });
+        connect(tasks, &Task::succeeded, [this, tasks]() {
+            QStringList warnings = tasks->warnings();
+            if (warnings.count())
+                CustomMessageBox::selectable(this, tr("Warnings"), warnings.join('\n'), QMessageBox::Warning)->show();
+
+            tasks->deleteLater();
+        });
+
+        for (auto& task : mdownload.getTasks()) {
+            tasks->addTask(task);
+        }
+
+        ProgressDialog loadDialog(this);
+        loadDialog.setSkipButton(true, tr("Abort"));
+        loadDialog.execWithTask(tasks);
+
+        m_model->update();
+    }
```

Comment:
This needs to be updated to reflect the fix I recently pushed to prevent use after free in #3670

Reply:
I think I will go with
![image](https://github.com/user-attachments/assets/526fd987-4a20-4afb-8b06-9afe2059d430)

Reply:
do not start the ~~merge~~rebase yet wait a few minutes for https://github.com/PrismLauncher/PrismLauncher/pull/3677 it will be merged into dev once the action finish

Reply:
you can do the merge now :)

Reply:
I already did the merge before you posted that. but no conflicts luckily!